File: frontend/src/perspective.js
// perspective.js — Frame Atlas V32
//
// Four-point perspective correction for crop mode: the maths for turning a
// tilted quadrilateral (a monitor, a poster, a projected image shot from an
// angle) back into a straight rectangle, plus the live preview that shows
// Ryan what the server is about to produce.
//
// THIS IS NOT PART OF THE DETECTION ENGINE. cropDetectV2.js decides where an
// axis-aligned rectangle goes and is untouched by any of this. Nothing here
// runs unless perspective mode is switched on, and perspective mode is never
// on by default.
//
// The corner points live in the image's natural (EXIF-upright) pixel space,
// exactly like cropDetectV2's boxes, and are converted to percentages only at
// the moment they're sent to the server — so the same four numbers mean the
// same selection whether the browser was looking at a 600px preview or a
// 6000px original.
//
// Corner order everywhere: [top-left, top-right, bottom-right, bottom-left].

export const CORNER_ORDER = ['tl', 'tr', 'br', 'bl'];

// Longest edge of the on-screen preview, in pixels. The preview is resampled
// in JS one pixel at a time, so this is a speed budget: it is redrawn on every
// pointermove while a corner is being dragged, and ~480px keeps that under a
// frame. The real output is computed server-side at full resolution.
const PREVIEW_MAX_EDGE = 480;

// Working copy of the source image the preview samples from. Full-resolution
// ImageData for a 6000x4000 photo is ~96MB, and we'd be re-reading it on every
// mouse move; a 1400px copy carries far more detail than a 480px preview can
// show and costs about 8MB.
const WORK_MAX_EDGE = 1400;

// One working copy per <img>, thrown away with the element.
const workCache = new WeakMap();

function getWorkPixels(img) {
  let work = workCache.get(img);
  if (work) return work;

  const nw = img.naturalWidth, nh = img.naturalHeight;
  const scale = Math.min(1, WORK_MAX_EDGE / Math.max(nw, nh));
  const w = Math.max(1, Math.round(nw * scale));
  const h = Math.max(1, Math.round(nh * scale));

  const canvas = document.createElement('canvas');
  canvas.width = w;
  canvas.height = h;
  const ctx = canvas.getContext('2d', { willReadFrequently: true });
  ctx.drawImage(img, 0, 0, w, h);
  const imageData = ctx.getImageData(0, 0, w, h);

  work = {
    w, h,
    // scale maps natural pixels -> working-copy pixels
    scale: w / nw,
    scaleY: h / nh,
    // Read as 32-bit words so one array index copies a whole pixel.
    px: new Uint32Array(imageData.data.buffer),
  };
  workCache.set(img, work);
  return work;
}

/**
 * Solve the 8 coefficients of the homography that maps each point in
 * `fromPts` onto the matching point in `toPts`.
 *
 *   X = (a*x + b*y + c) / (g*x + h*y + 1)
 *   Y = (d*x + e*y + f) / (g*x + h*y + 1)
 *
 * Multiplying out gives two linear equations per corner; four corners make an
 * 8x8 system. This is the same solve app.py does server-side (there in plain
 * Python, here in plain JS) so the preview and the real output agree.
 *
 * Returns [a,b,c,d,e,f,g,h], or null if the four points are degenerate.
 */
export function solveHomography(fromPts, toPts) {
  const m = [];
  const rhs = [];
  for (let i = 0; i < 4; i++) {
    const [x, y] = fromPts[i];
    const [tx, ty] = toPts[i];
    m.push([x, y, 1, 0, 0, 0, -tx * x, -tx * y]); rhs.push(tx);
    m.push([0, 0, 0, x, y, 1, -ty * x, -ty * y]); rhs.push(ty);
  }

  const n = 8;
  const a = m.map((row, i) => [...row, rhs[i]]);

  for (let col = 0; col < n; col++) {
    // Partial pivoting. The handles are seeded from a rectangle, so exactly
    // axis-aligned corners — which put a zero on the diagonal — are the
    // COMMON case here, not an exotic one.
    let pivot = col;
    for (let r = col + 1; r < n; r++) {
      if (Math.abs(a[r][col]) > Math.abs(a[pivot][col])) pivot = r;
    }
    if (Math.abs(a[pivot][col]) < 1e-12) return null;
    [a[col], a[pivot]] = [a[pivot], a[col]];

    const inv = 1 / a[col][col];
    for (let r = col + 1; r < n; r++) {
      const factor = a[r][col] * inv;
      if (factor === 0) continue;
      for (let k = col; k <= n; k++) a[r][k] -= factor * a[col][k];
    }
  }

  const out = new Array(n).fill(0);
  for (let row = n - 1; row >= 0; row--) {
    let acc = a[row][n];
    for (let k = row + 1; k < n; k++) acc -= a[row][k] * out[k];
    out[row] = acc / a[row][row];
  }
  return out;
}

const dist = (p, q) => Math.hypot(p.x - q.x, p.y - q.y);

/**
 * Pixel size of the de-skewed rectangle: each side is the AVERAGE of the two
 * OPPOSITE edges of the quad. Identical rule to perspective_output_size() in
 * app.py — sizing off a single edge would bake the perspective back in
 * (the near edge of an angled monitor is longer than the far one, so it would
 * stretch the whole result).
 */
export function quadOutputSize(corners) {
  const [tl, tr, br, bl] = corners;
  return {
    w: Math.round((dist(tl, tr) + dist(bl, br)) / 2),
    h: Math.round((dist(tl, bl) + dist(tr, br)) / 2),
  };
}

/** Four corners at the untouched image bounds — nothing to correct. */
export function isIdentityQuad(corners, img) {
  if (!corners || !img) return true;
  const W = img.naturalWidth, H = img.naturalHeight;
  const want = [[0, 0], [W, 0], [W, H], [0, H]];
  // Half a percent of each dimension, matching the server's tolerance.
  const tx = W * 0.005, ty = H * 0.005;
  return corners.every((p, i) =>
    Math.abs(p.x - want[i][0]) <= tx && Math.abs(p.y - want[i][1]) <= ty);
}

/**
 * Is this a simple (non-crossing) convex quadrilateral, in the given order?
 *
 * Same test the server runs, for the same reason: a rectangle photographed
 * from ANY angle is always convex, so a crossed ("bow-tie") or dented outline
 * is not a perspective view of anything. The server refuses these outright —
 * running the check here too means Ryan sees the outline turn red while he's
 * dragging instead of getting a rejection after he hits Apply.
 */
export function isConvexQuad(corners) {
  if (!corners || corners.length !== 4) return false;
  let sign = 0;
  for (let i = 0; i < 4; i++) {
    const a = corners[i], b = corners[(i + 1) % 4], c = corners[(i + 2) % 4];
    const cross = (b.x - a.x) * (c.y - b.y) - (b.y - a.y) * (c.x - b.x);
    if (cross === 0) return false;
    const s = cross > 0 ? 1 : -1;
    if (sign === 0) sign = s;
    else if (s !== sign) return false;
  }
  return true;
}

/** Seed four corners from an axis-aligned box (or the whole image). */
export function cornersFromBox(box, img) {
  const b = box || { x: 0, y: 0, w: img.naturalWidth, h: img.naturalHeight };
  return [
    { x: b.x,       y: b.y },
    { x: b.x + b.w, y: b.y },
    { x: b.x + b.w, y: b.y + b.h },
    { x: b.x,       y: b.y + b.h },
  ];
}

/** Corners in natural pixels -> the percentages the API expects. */
export function cornersToPercent(corners, img) {
  const W = img.naturalWidth, H = img.naturalHeight;
  // Clamped to 0-100 because the server REJECTS out-of-range corners rather
  // than clamping them (clamping would silently change the shape). The drag
  // handler already keeps them in bounds; this only guards floating-point
  // dust like 100.00000000000001.
  return corners.map(p => ({
    x: Math.min(100, Math.max(0, (p.x / W) * 100)),
    y: Math.min(100, Math.max(0, (p.y / H) * 100)),
  }));
}

/**
 * Draw the de-skewed result into `canvas`, so the CROPPED RESULT panel shows
 * the actual perspective correction rather than a rectangle that ignores it.
 *
 * Nearest-neighbour on purpose: this is a preview redrawn on every
 * pointermove. The server uses bicubic on the full-resolution original.
 */
export function drawPerspectivePreview(canvas, img, corners) {
  const out = quadOutputSize(corners);
  if (!(out.w > 0 && out.h > 0)) return false;

  const fit = Math.min(1, PREVIEW_MAX_EDGE / Math.max(out.w, out.h));
  const dw = Math.max(1, Math.round(out.w * fit));
  const dh = Math.max(1, Math.round(out.h * fit));

  const work = getWorkPixels(img);

  // Same direction as Pillow: walk the OUTPUT pixels and ask where each one
  // came from. So the homography maps the destination rectangle onto the quad,
  // pre-scaled into working-copy coordinates.
  const dst = [[0, 0], [dw, 0], [dw, dh], [0, dh]];
  const src = corners.map(p => [p.x * work.scale, p.y * work.scaleY]);
  const co = solveHomography(dst, src);
  if (!co) return false;
  const [a, b, c, d, e, f, g, h] = co;

  canvas.width = dw;
  canvas.height = dh;
  const ctx = canvas.getContext('2d');
  const outData = ctx.createImageData(dw, dh);
  const outPx = new Uint32Array(outData.data.buffer);

  for (let y = 0; y < dh; y++) {
    for (let x = 0; x < dw; x++) {
      const w = g * x + h * y + 1;
      const sx = (a * x + b * y + c) / w;
      const sy = (d * x + e * y + f) / w;
      const ix = sx | 0, iy = sy | 0;
      // Outside the source reads as transparent. The quad is constrained to
      // the image bounds, so in practice this only catches the last row/column
      // of rounding at the very edge.
      outPx[y * dw + x] =
        (ix >= 0 && iy >= 0 && ix < work.w && iy < work.h)
          ? work.px[iy * work.w + ix]
          : 0;
    }
  }

  ctx.putImageData(outData, 0, 0);
  return true;
}

File: backend/perspective.py
"""Frame Atlas — perspective crop geometry (V32).

Split out of app.py in V45 (Day 27). Pure geometry: percentages and PIL images
in, coefficients and a straightened image out. No database, no Drive, no Flask.

Ryan photographs monitors, posters and projected images that sit on an angle.
A rectangle can't extract those, so crop gained a second SHAPE — four
independently-dragged corners, straightened back into a rectangle. It is not a
second crop tool: both shapes converge before the destructive Drive write in
app.py, so the backup-first rule and the post-write refresh are the same lines
for both. Never fork that tail (V27 crashed after overwriting Drive and left
the database holding a pre-crop thumbnail).

The 8 homography coefficients are solved by hand here and again in
frontend/src/perspective.js, cross-checked to 9 decimals. numpy is
deliberately NOT a dependency — it isn't in backend/requirements.txt, and
touching that file adds 3+ minutes to every Railway deploy.
"""

import math

from PIL import Image

# The quad has to cover at least this much of the frame (percent of the whole
# image area). Same spirit as the rectangle path's "w >= 1% and h >= 1%":
# a stray double-click that collapses the quad to a speck should be refused
# with a readable message, not turned into an 8-pixel file in Drive.
PERSPECTIVE_MIN_AREA_PCT = 1.0

# Two corners closer together than this (in percent-of-frame units) are
# treated as the same point — a degenerate "triangle" quad, which has no
# unique perspective solution and makes the 8x8 solve singular.
PERSPECTIVE_MIN_CORNER_GAP_PCT = 0.5


def _quad_signed_area(pts):
    """Shoelace area of a polygon. The sign tells you which way the points
    wind; the magnitude is the area in whatever units the points are in."""
    total = 0.0
    n = len(pts)
    for i in range(n):
        x1, y1 = pts[i]
        x2, y2 = pts[(i + 1) % n]
        total += x1 * y2 - x2 * y1
    return total / 2.0


def parse_perspective_corners(raw):
    """Validate the four corner points the browser sends and hand back
    [(x, y), ...] in percent (0-100), in the order the caller gave them:
    top-left, top-right, bottom-right, bottom-left.

    Raises ValueError with a message written for Ryan, not for a log file.

    Why this is stricter than the rectangle path, which just clamps: clamping
    a rectangle is harmless — it stays a rectangle in the same place. Clamping
    ONE corner of a quadrilateral silently changes the SHAPE of the correction,
    so the file Drive ends up holding would not be the one the preview showed.
    Every check below has to fail HERE, before the destructive write, because a
    crop overwrites the Drive file in place and there is no undo.
    """
    if not isinstance(raw, (list, tuple)) or len(raw) != 4:
        raise ValueError('Perspective correction needs exactly four corner points.')

    pts = []
    for p in raw:
        if isinstance(p, dict):
            cx, cy = p.get('x'), p.get('y')
        elif isinstance(p, (list, tuple)) and len(p) == 2:
            cx, cy = p
        else:
            raise ValueError('Each corner must be a point with numeric x and y percentages.')
        try:
            cx = float(cx)
            cy = float(cy)
        except (TypeError, ValueError):
            raise ValueError('Each corner must be a point with numeric x and y percentages.')
        if not (math.isfinite(cx) and math.isfinite(cy)):
            raise ValueError('Corner points must be real numbers.')
        # Rejected, not clamped — see the docstring.
        if not (0.0 <= cx <= 100.0) or not (0.0 <= cy <= 100.0):
            raise ValueError('Corner points must sit inside the image (0-100%).')
        pts.append((cx, cy))

    # Two corners on top of each other collapse the quad to a triangle, which
    # has no unique perspective solution (the 8x8 system below goes singular).
    for i in range(4):
        for j in range(i + 1, 4):
            if math.dist(pts[i], pts[j]) < PERSPECTIVE_MIN_CORNER_GAP_PCT:
                raise ValueError('Two corner points are on top of each other — drag them apart.')

    # Convexity, checked as "every turn goes the same way". This is the
    # bow-tie guard: drag the top-left handle past the top-right one and the
    # edges cross, one cross product flips sign, and we refuse. It also
    # rejects concave quads, which is correct rather than merely convenient —
    # a rectangle photographed from ANY angle is always convex, so a dented
    # quad is not a perspective view of anything and the transform would fold
    # the picture over itself.
    signs = []
    for i in range(4):
        ax, ay = pts[i]
        bx, by = pts[(i + 1) % 4]
        cx, cy = pts[(i + 2) % 4]
        cross = (bx - ax) * (cy - by) - (by - ay) * (cx - bx)
        if abs(cross) < 1e-9:
            raise ValueError('Those four corners are in a straight line — they do not enclose an area.')
        signs.append(cross > 0)
    if len(set(signs)) != 1:
        raise ValueError('Those corners cross over each other. Drag them so the outline is a simple four-sided shape.')

    # Winding direction is deliberately NOT pinned. A flipped quad produces a
    # mirrored result, which is exactly what the on-screen preview shows, so
    # it stays the user's call rather than a confusing rejection.

    if abs(_quad_signed_area(pts)) < PERSPECTIVE_MIN_AREA_PCT:
        raise ValueError('That selection is too small — it must cover at least 1% of the image.')

    return pts


def perspective_is_whole_image(pts, tolerance=0.5):
    """True when the quad is (within a hair of) the untouched image corners —
    a no-op. Mirrors the rectangle path's "covers the whole image" refusal so
    a stray Apply can't burn a destructive Drive rewrite doing nothing."""
    want = [(0.0, 0.0), (100.0, 0.0), (100.0, 100.0), (0.0, 100.0)]
    return all(abs(p[0] - q[0]) <= tolerance and abs(p[1] - q[1]) <= tolerance
               for p, q in zip(pts, want))


def solve_linear_system(matrix, rhs):
    """Gaussian elimination with partial pivoting. Returns the solution vector.

    Deliberately hand-written: numpy would do this in one line, but numpy is
    NOT in requirements.txt, and adding it costs every Railway deploy an extra
    few minutes forever in exchange for one 8x8 solve that fits in 20 lines.
    """
    n = len(rhs)
    # Work on copies — callers reuse their coefficient rows.
    a = [list(row) + [rhs[i]] for i, row in enumerate(matrix)]

    for col in range(n):
        # Partial pivoting: swap the largest remaining magnitude into place.
        # Without it a perfectly legitimate quad whose diagonal entry happens
        # to be zero — an exactly axis-aligned corner, which is the COMMON
        # case here since the handles are seeded from a rectangle — divides
        # by zero instead of solving.
        pivot = max(range(col, n), key=lambda r: abs(a[r][col]))
        if abs(a[pivot][col]) < 1e-12:
            raise ValueError('Those four corners do not describe a valid perspective.')
        a[col], a[pivot] = a[pivot], a[col]

        inv = 1.0 / a[col][col]
        for r in range(col + 1, n):
            factor = a[r][col] * inv
            if factor == 0.0:
                continue
            for k in range(col, n + 1):
                a[r][k] -= factor * a[col][k]

    out = [0.0] * n
    for row in range(n - 1, -1, -1):
        acc = a[row][n] - sum(a[row][k] * out[k] for k in range(row + 1, n))
        out[row] = acc / a[row][row]
    return out


def solve_perspective_coeffs(from_pts, to_pts):
    """The eight numbers Pillow's Image.PERSPECTIVE wants.

    A perspective map is  X = (a*x + b*y + c) / (g*x + h*y + 1)
                          Y = (d*x + e*y + f) / (g*x + h*y + 1)
    Multiplying out gives two linear equations per corner; four corners make
    an 8x8 system in (a, b, c, d, e, f, g, h).

    IMPORTANT DIRECTION NOTE: Pillow's transform walks the OUTPUT pixels and
    asks where each one came FROM, so the coefficients must map
    output -> input. Call this with from_pts = the destination rectangle and
    to_pts = the quad in the source photo. Getting this backwards warps the
    picture the wrong way instead of raising, so it is worth the paragraph.
    """
    matrix = []
    rhs = []
    for (x, y), (tx, ty) in zip(from_pts, to_pts):
        matrix.append([x, y, 1, 0, 0, 0, -tx * x, -tx * y])
        rhs.append(tx)
        matrix.append([0, 0, 0, x, y, 1, -ty * x, -ty * y])
        rhs.append(ty)
    return solve_linear_system(matrix, rhs)


def perspective_output_size(src_px):
    """Pixel size of the de-skewed rectangle, derived from the quad's own edges.

    Each side of the output is the AVERAGE of the two OPPOSITE edges of the
    quad. Picking one edge instead would bake the perspective back into the
    result: on a monitor shot from the left, the near (right) edge is much
    longer than the far one, so sizing off the near edge stretches the whole
    image and sizing off the far edge squashes it. The average lands between
    the two and keeps the subject's proportions roughly true, which is what
    "de-skew" should mean. It also can't invent detail — the output is never
    larger than the longest edge the photo actually contains.
    """
    tl, tr, br, bl = src_px
    top = math.dist(tl, tr)
    bottom = math.dist(bl, br)
    left = math.dist(tl, bl)
    right = math.dist(tr, br)
    return int(round((top + bottom) / 2.0)), int(round((left + right) / 2.0))


def perspective_correct(img, corners_pct):
    """Straighten the quadrilateral `corners_pct` (percentages, TL/TR/BR/BL)
    out of `img` into a plain rectangle. Returns a new PIL image."""
    w_px, h_px = img.width, img.height
    # Percent -> pixels happens HERE and only here, which is what lets the same
    # four numbers mean the same selection whether the browser was looking at a
    # 600px preview or the 6000px original.
    src = [(x / 100.0 * w_px, y / 100.0 * h_px) for x, y in corners_pct]

    out_w, out_h = perspective_output_size(src)
    # Same floor the rectangle path uses — below this there is no usable
    # picture left and the downstream thumbnail/palette work gets silly.
    if out_w < 8 or out_h < 8:
        raise ValueError('Perspective selection is too small to produce a usable image')

    dst = [(0, 0), (out_w, 0), (out_w, out_h), (0, out_h)]
    coeffs = solve_perspective_coeffs(dst, src)
    return img.transform((out_w, out_h), Image.PERSPECTIVE, coeffs, resample=Image.BICUBIC)
